copy_paste: improve multi message selection and paste format.

apoorvapendse opened the pull request:
Summary:
Earlier, selecting multiple messages led to copying the entire content of those messages.
This applied even for the partially selected messages at the start/end of the selection.
The selection wasn't updated to reflect this.

This fix deals with this confusing behavior in two ways, depending on the browser.
1. On Chrome, only the content that is selected is copied for a multi-message selection.
2. On Firefox, the copy behavior is same as before, but the selection expands to show the user what they will be actually copying. This reduces the mismatch between what is selected and what is copied.

Although 1 is the better solution here, implementing this on Firefox using ranges felt tricky and unreliable to me.
Some observations are noted at: [#design > Improve the message copying experience #6316 @ 💬](https://chat.zulip.org/#narrow/channel/101-design/topic/Improve.20the.20message.20copying.20experience.20.236316/near/2210889)

Fixes: #6316.
CZO: [#design > Improve the message copying experience #6316](https://chat.zulip.org/#narrow/channel/101-design/topic/Improve.20the.20message.20copying.20experience.20.236316/with/2192405)



| Firefox|
|-----|
|![firefox](https://github.com/user-attachments/assets/49ca1315-70ce-4840-8012-65865f597f9f)|


|Chrome|
|------|
|![chrome](https://github.com/user-attachments/assets/e97c62a9-565d-4616-998a-2dcfcd149c14)|


Edit: 3 Jun 2026.
This got another mention in rust-lang Zulip.
[#zulip > Copying the text I selected](https://rust-lang.zulipchat.com/#narrow/channel/122653-zulip/topic/Copying.20the.20text.20I.20selected/with/599577234)

apoorvapendse committed:
copy_messages: Improve multi message selection.

Earlier, selecting multiple messages led to copying the
entire content of those messages.
This applied even for the partially selected messages
at the start/end of the selection.
The selection wasn't updated to reflect this.

This fix deals with this confusing behavior in two ways,
depending on the browser.

1. On Chrome, which currently maintains a single range
   for a multi-message selection only the content that
   is selected is copied for a multi-message selection.
2. On Firefox, whose baseline currently keeps multiple ranges for
   `user-select:none` elements, the copy behavior is same as before,
   but the selection expands to show the user what they will be actually
   copying.

This reduces the mismatch between what is selected and what is copied.

Although 1 is the better solution here, implementing this on Firefox
using ranges felt tricky and unreliable to me.

Firefox 147 has the fix for the quirk where selection over
non-selectable elements doesn't create multiple ranges.
We should remember to throw away the multi-range handling logic
when that ver. becomes the baseline.

discussion: https://github.com/zulip/zulip/pull/35100#issuecomment-4683858639

Fixes: #6316.

CZO:
https://chat.zulip.org/#narrow/channel/101-design/topic/Improve.20the.20message.20copying.20experience.20.236316/with/2192405.
Signed-off-by: apoorvapendse <apoorvavpendse@gmail.com>

whilstsomebody commented:
@apoorvapendse, could you add the previous behavior for each browsers? It might not be easy to understand otherwise what we are actually doing in the PR.

whilstsomebody reviewed (approved):
Okh, I got it now after playiing a bit. 
Looks good in manual testing and looked into the code as well which also looks fine.
Passing it for now as mentors and maintainers might have better opinion on the code.

apoorvapendse commented:
@zulipbot add "mentor review"
Tagging @punchagan, @vinitS101 for mentor review.

@zulipbot remove "buddy review"

vinitS101 commented:
Looks good to me

apoorvapendse commented:
@zulipbot remove "mentor review"

alya commented:
In Chrome, copying part of one bullet in a bulleted list copies the entire message for me.

alya commented:
Maybe you can test with other special formatting like numbered lists, spoilers, etc.

apoorvapendse commented:
I've addressed the bug @alya found:
https://github.com/user-attachments/assets/15106da8-963e-4f4d-95a0-ed039907a727


https://github.com/user-attachments/assets/14aa3b12-f014-4765-b86d-7b9ed38e0ffe


https://github.com/user-attachments/assets/96b6f710-2ceb-45a7-bc87-074b7bd9cff1

alya commented:
@amanagr Would you be up for being the next reviewer on this one?

amanagr reviewed (commented):
These changes work for me when I tested in Firefox and Chrome on Ubuntu.

Left a couple of comments, otherwise looks good to me.

apoorvapendse commented on web/src/copy_messages.ts:
```diff
@@ -280,6 +335,22 @@ export function copy_handler(ev: ClipboardEvent): boolean {
+    // We do this only for Firefox where multi-message selections are
+    // broken down into multiple-ranges.
+    // For details, see: https://chat.zulip.org/#narrow/channel/101-design/topic/Improve.20the.20message.20copying.20experience.20.236316/near/2210892
+    const first = $(`[data-message-id='${start_id}']`).find(".messagebox-content")[0];
```
`.focused-message-list`
Is this necessary in the selector?
I tried constructing a few narrows, and always end up seeing only one `.message-list` along with this class in the DOM.

amanagr replied:
We don't destroy `message-list` for `All messages` if  it is your home view, so, it is possible to have multiple. Also, in the future we don't which direction we will go.

amanagr commented:
Changes look good to me. Tested manually too.

amanagr commented:
@apoorvapendse can you look into a related bug that copying message in logged out view show the login modal?

apoorvapendse commented:
It looks like @opmkumar has a fix up for this #35847

timabbott commented on web/src/copy_messages.ts:
```diff
@@ -52,6 +53,48 @@ function construct_recipient_header($message_row: JQuery): JQuery {
+        if (navigator.userAgent.includes("Chrome")) {
+            blueslip.error("Multiple ranges detected in Chrome for a multi-message selection.");
+        }
+        return {$first_element, $last_element};
```
These are both still `null` here; why is that correct behavior?

timabbott commented on web/src/copy_messages.ts:
```diff
@@ -52,6 +53,48 @@ function construct_recipient_header($message_row: JQuery): JQuery {
     return $("<p>").append($("<strong>").text(message_header_content));
 }
+
+function get_first_and_last_selected_elements(): {
```
This name is confusing, in that it seems to never return an actual element in the message, but instead something constructed virtually?

timabbott replied:
Actually this abstraction seems weird; it seems like the main determination this is making is the determination of the `$fragment_children` elements; would it make sense to just do that logic separately for the start and end of the selection?

apoorvapendse commented on web/src/copy_messages.ts:
```diff
@@ -93,7 +145,14 @@ function construct_copy_div($div: JQuery, start_id: number, end_id: number): voi
+            } else {
+                $content = $last_element;
+            }
+        }
```
Right, added a test and used elif to avoid `$(message.content)`

apoorvapendse reviewed (commented):
There were a couple edge cases in terms of the message header being selected before the first message(or after the last message) on Firefox and removed from the selection as a part of the selection update logic.
We now preserve that in case the the first range or last range contents aren't inside a `.message_row`.

```ts
  // We only want to expand the selection ranges if start/end range lies within a message.
    // Not having these checks could alter the range in weird ways, like a message header
    // selection getting removed to select the first message that follows that header.
    if (
        get_nearest_html_element(window.getSelection()!.getRangeAt(0).startContainer)!.closest(
            ".message_row",
        )
    ) {
        const $start = copy_rows[0];
        assert($start?.[0] !== undefined);
        const start_node = $start[0];
        window.getSelection()?.getRangeAt(0).setStartBefore(start_node);
    }
 ```

alya commented:
@amanagr Looks like this one is on your plate again. :)

apoorvapendse committed:
compose_paste: Update regex to match only list elements.

The previous version incorrectly matched with and
replaced parts of strings like "\n\n\n**hello**".

Signed-off-by: apoorvapendse <apoorvavpendse@gmail.com>

apoorvapendse committed:
copy_paste: Update formating for multi-message paste.

Changes:
- Two newlines are inserted before every message header,
except the first.
- The message recipient text uses bold text followed by "| <date>".
- A newline is inserted after a message header, before the sender
name.
- The sender name uses bold text.
- A newline is inserted after the sender name, before the
message content.
- Two newlines separate consecutive messages from the same
recipient, so each message's body and the next message's sender
name are split by a blank line.

Discussion:
https://chat.zulip.org/#narrow/channel/101-design/topic/Multiple.20message.20paste.20formatting/with/2333459.
Signed-off-by: apoorvapendse <apoorvavpendse@gmail.com>
Co-authored-by: Shubham Padia <shubham@zulip.com>

apoorvapendse commented:
Updated the PR to update the formatting, as per the disccusion on [#design > Multiple message paste formatting](https://chat.zulip.org/#narrow/channel/101-design/topic/Multiple.20message.20paste.20formatting/with/2333459)

<details>
<summary> screenshots</summary>

Paste output for messages having different recipients:
<img width="613" height="539" alt="image" src="https://github.com/user-attachments/assets/d5ed53ef-a54c-49ef-a15a-3e339e8a1ea1" />

Paste output on Google docs:
<img width="840" height="418" alt="image" src="https://github.com/user-attachments/assets/db33430d-034e-4a75-be46-90267f7ebc69" />

Paste output for messages having same recipient:
<img width="601" height="256" alt="image" src="https://github.com/user-attachments/assets/070d5bf5-2f97-4cc4-83c3-7f3e90b4e897" />

Paste output for input containing `/me` messages:
<img width="793" height="549" alt="image" src="https://github.com/user-attachments/assets/33f95e7c-8b63-423e-b39c-f8745d8a639c" />


</details>

apoorvapendse commented on web/tests/compose_paste.test.cjs:
```diff
@@ -386,6 +386,14 @@ run_test("paste_handler_converter", () => {
+        compose_paste.paste_handler_converter(input),
+        "**test > LAST API has been skipping erratically** | Dec 20\n**Othello, the Moor of Venice:**\nWhen shall we three meet again In thunder, lightning, or in rain?\n\n**test > leased keurig wasn't loading** | Dec 21\n**King Hamlet:**\nSit down awhile; And let us once again assail your ears, That are so fortified against our story What we have two nights seen.",
+    );
+
```
This test also indirectly tests the regex change introduced in the previous commit.

apoorvapendse committed:
copy: Append/prepend ellipsis when partially selecting messages.

Signed-off-by: apoorvapendse <apoorvavpendse@gmail.com>

apoorvapendse commented:
This is ready for another look.

alya commented:
@amanagr Can you tag for integration review when ready?

@timabbott The screenshots look reasonable to me; let me know if you'd like me to re-test.

apoorvapendse commented:
Gentle bump for @timabbott

timabbott commented:
@apoorvapendse is this one still current? I wasn't sure given the recent work on multi-message quoting how this relates, since it doesn't obviously share code with those changes.

apoorvapendse commented:
Yeah, it predates the multi-message quote PR.
It's safe to integrate it, I'll look into doing partial quoting in the multi message quote section later on.

Opened #39129 to track that.

apoorvapendse commented:
@amanagr  / @laurynmm can we do an integration review on this?

emilysutherlin commented:
Note to self to make sure we have the folks in the[ Rust community conversation](https://rust-lang.zulipchat.com/#narrow/channel/122653-zulip/topic/Copying.20the.20text.20I.20selected/with/599876326) topic test this for screenreader use once it's released to Cloud.

amanagr commented:
I have already done one round of review on this, so it wouldn't make sense for me to do another round. @laurynmm already has a big queue, so I would ask @shubham-padia for a review since he has the most recent merged work in the compose space.

shubham-padia reviewed (commented):
I haven't looked deeply at the code, but the Firefox claim does not hold true anymore: https://www.mail-archive.com/dev-platform@mozilla.org/msg01659.html (Claude helped me find this). Next step would be to check whether the browser behaviour we need is supported by the browsers we support, see our browserslist. We can chose to discard the code for the alternate behaviour after checking this.

Beyond that, see this selection bug that happens when the sender name is selected as the last part of the selection.

<img width="926" height="634" alt="Image" src="https://github.com/user-attachments/assets/63f83e8c-219b-4365-ada5-91a8ba4a882b" />

apoorvapendse commented:
This is so cool, it's gonna end up vastly simplifying the the copy code once it becomes the baseline!

> Next step would be to check whether the browser behaviour we need is supported by the browsers we support, see our browserslist.

browserslist points to FF 120+
So I think we can file an issue to remember to remove the firefox multi-range logic in the future?

as seen here: https://bugzilla.mozilla.org/show_bug.cgi?id=1998858 , it seems to have been resolved in FF 147.

Also, fixed that bug you reported.
It was because how `analyze_selection` and `find_boundary_tr` picks the end_id, because it only looks at `.messagebox-content`'s presence in the range's end container and `.messagebox-content` parents `.message_sender` and friends.

shubham-padia commented on web/src/copy_messages.ts:
```diff
@@ -52,6 +54,94 @@ function construct_recipient_header($message_row: JQuery): JQuery {
+// anyways contain the entire `.message_content` HTML.
+function get_selected_message_content_html_at(
+    idx: number,
+    selected_message_content_elements: NodeListOf<HTMLElement> | undefined,
```
idx here seems confusing. Just looking at the function, I would assume idx would be any number, while we only ever pass 0 or -1. We should use some other way to indicate first and last.

Once that is fixed, we might want to rethink the name of the function and see if `_at` fits there anymore or not.

shubham-padia replied:
If we're doing our own processing, that should be somehow indicated in the function name so it doesn't feel like we are getting exactly what the user has selected.

apoorvapendse replied:
Yeah, but I think we might benefit by supporting any `idx` along with first/last when we implement [#design > select messages to move/delete](https://chat.zulip.org/#narrow/channel/101-design/topic/select.20messages.20to.20move.2Fdelete/with/2184173) and want to support bulk, discontiguous copy there.

I don't think it hurts to keep this flexible?
I'll try to think of a name for the second comment.

shubham-padia replied:
The concern with the idx being there is that when reading the function by it's own, without deep context about this part of the codebase (which I don't), I would think any idx is being passed here, when only 0 and -1 are being actually used. Mentioning first/last explicitly keeps the function easy to understand right now and later switching to an index when need be should not be difficult in the future.

apoorvapendse replied:
updated


<details>
<summary> diff </summary>

```diff
 // The caller is expected to only pass the first or last message
 // from a selection range, as the intermediate selected messages
 // anyways contain the entire `.message_content` HTML.
-function get_selected_message_content_html_at(
-    idx: number,
+function get_html_for_bookend_message_content(
+    type: RangeContainer,
     selected_message_content_elements: NodeListOf<HTMLElement> | undefined,
 ): string {
     assert(window.getSelection()?.rangeCount === 1);
@@ -80,13 +82,11 @@ function get_selected_message_content_html_at(
     // We don't want to append the same content as the first message in the selection
     // if we are trying to get the `.message_content` HTML for the last
     // message when there is only one `.message_content` in the selected range.
-    if (selected_message_content_elements.length === 1 && idx < 0) {
+    if (selected_message_content_elements.length === 1 && type === "end") {
         return "";
     }
 
-    if (idx < 0) {
-        idx = selected_message_content_elements.length - 1;
-    }
+    const idx = type === "start" ? 0 : selected_message_content_elements.length - 1;
     const target_message_content_element = selected_message_content_elements[idx];
     assert(target_message_content_element !== undefined);
 
@@ -100,6 +100,17 @@ function get_selected_message_content_html_at(
     return target_message_content_element.innerHTML;
 }
```

</details>

apoorvapendse replied:
Thinking a bit harder, I guess I was hallucinating yesterday.

Here is a comment that I wrote:
```
// The caller is expected to only pass the first or last message
// from a selection range, as the intermediate selected messages
// anyways contain the entire `.message_content` HTML.
```

Partial selection would probably never be possible the way I spoke in https://github.com/zulip/zulip/pull/35100/changes/b36612a8814035396ae276d819f96755f7be4eab#r3413368676

shubham-padia commented on web/src/copy_messages.ts:
```diff
@@ -52,6 +54,94 @@ function construct_recipient_header($message_row: JQuery): JQuery {
+
+    if (navigator.userAgent.includes("Chrome")) {
+        blueslip.error("Multiple ranges detected in Chrome for a multi-message selection.");
+    }
```
Why are we raising an error on this?

apoorvapendse replied:
This was discussed [here](https://github.com/zulip/zulip/pull/35100#discussion_r2284617810).
It would serve as a natural reminder to toggle the multi-range code in case something weird like Chrome switching to multiple ranges happens in the future, I think.

shubham-padia replied:
Understood!

shubham-padia commented on web/src/copy_messages.ts:
```diff
@@ -52,6 +54,94 @@ function construct_recipient_header($message_row: JQuery): JQuery {
+    const is_end_within_message_row =
+        get_nearest_html_element(
+            window.getSelection()!.getRangeAt(range_count - 1).endContainer,
+        )!.closest(".message_row") !== null;
```
Feels like too much stuff going on in a single line and it becomes hard to read, maybe we can have something along the lines of

```
  function is_container_within_message_row(node: Node): boolean {
      return get_nearest_html_element(node)?.closest(".message_row") != null;
  }
```

just to make this read better. Any improvements beyond that are also welcome, this was just an initial idea.

apoorvapendse replied:
ok, changed it to:
<details>
<summary> diff </summary>

```diff
diff --git a/web/src/copy_messages.ts b/web/src/copy_messages.ts
index b293b75630..ab88e18cb2 100644
--- a/web/src/copy_messages.ts
+++ b/web/src/copy_messages.ts
@@ -9,6 +9,8 @@ import * as message_lists from "./message_lists.ts";
 import * as rows from "./rows.ts";
 import {the} from "./util.ts";
 
+type RangeContainer = "start" | "end";
+
 function find_boundary_tr(
     $initial_tr: JQuery,
     iterate_row: ($tr: JQuery) => JQuery,
@@ -100,6 +102,17 @@ function get_selected_message_content_html_at(
     return target_message_content_element.innerHTML;
 }
 
+function is_container_within_message_row(type: RangeContainer) : boolean {
+    const range_count = window.getSelection()?.rangeCount;
+    assert(range_count && range_count > 1);
+
+    const target_range_idx = type === 'start' ? 0 : range_count - 1;
+    const range = window.getSelection()!.getRangeAt(target_range_idx);
+    const container = type === 'start' ? range.startContainer : range.endContainer
+    assert(container !== undefined);
+    return get_nearest_html_element(container)?.closest(".message_row") !== null;
+}
+
 function maybe_expand_selection_for_first_and_last_messages(
     copy_rows: JQuery[],
     range_count: number,
@@ -115,24 +128,14 @@ function maybe_expand_selection_for_first_and_last_messages(
     // We only want to expand the selection ranges if start/end range lies within a message.
     // Not having these checks could alter the range in weird ways, like a message header
     // selection getting removed to select the first message that follows that header.
-    const is_start_within_message_row =
-        get_nearest_html_element(window.getSelection()!.getRangeAt(0).startContainer)!.closest(
-            ".message_row",
-        ) !== null;
-
-    const is_end_within_message_row =
-        get_nearest_html_element(
-            window.getSelection()!.getRangeAt(range_count - 1).endContainer,
-        )!.closest(".message_row") !== null;
-
-    if (is_start_within_message_row) {
+    if (is_container_within_message_row("start")) {
         const $start = copy_rows[0];
         assert($start?.[0] !== undefined);
         const start_node = the($start.find(".message_content"));
         window.getSelection()?.getRangeAt(0).setStartBefore(start_node);
     }
 
-    if (is_end_within_message_row) {
+    if (is_container_within_message_row("end")) {
         const $end = copy_rows.at(-1);
         assert($end?.[0] !== undefined);
         const end_node = the($end.find(".message_content"));
```

</details>

shubham-padia commented on web/src/copy_messages.ts:
```diff
@@ -71,7 +161,59 @@ function construct_copy_div($div: JQuery, start_id: number, end_id: number): voi
+    } else {
+        // Instead of copying the entire content of the first and last message,
+        // we only use the content that is part of the selection.
+        // This is only done on Chrome for now, because of the behavior of having
```
Across the entire PR, instead of using chrome vs firefox, we should state browsers that have multi range selection and browsers that do not, whenever we are mentioning this issue. At one place (not sure where, but where most of this logic lives, so most probably this place), we should write that `At the time of writing (date), chrome supports this and firefox that`, mention a TODO of removing the multi range thing once fulfill our browser baseline

apoorvapendse replied:
Hmm, I party disagree with this.
I think the invisible contract was that we only care about Chrome and FF, cause other browsers are generally just spinoffs over these, except Safari.

Not sure what we gain by not keeping observations explicit.
Whenever I've had Tim review PRs in this area over the past couple years, the conversation was always about not breaking the behaviour in these two and rest should either fall in place or is too niche to care about.

shubham-padia replied:
Understood, we can let this be as-is then.

shubham-padia commented on web/src/copy_messages.ts:
```diff
@@ -71,7 +161,59 @@ function construct_copy_div($div: JQuery, start_id: number, end_id: number): voi
+            // message, even though there is no meaningful `.message_content` selected for
+            // that message.
+            // So the actually  last message of this selection is
+            // at copy_rows[copy_rows.length - 2]
```
This comment does not explain what is going on here at a quick glance. We want to highlight what this behaviour is more than what condition we are addressing technically. If we are solving for selecting only (entire or some of) the name at the end, we should highlight that scenario.

apoorvapendse replied:
Updated.

shubham-padia commented on web/src/copy_messages.ts:
```diff
@@ -71,7 +161,59 @@ function construct_copy_div($div: JQuery, start_id: number, end_id: number): voi
+        // a single range for a multi-message selection.
+        const selected_message_content_elements = get_selected_message_content_elements();
+        assert(selected_message_content_elements !== undefined);
+        if (selected_message_content_elements.length === copy_rows.length - 1) {
```
A comment on top of this, something that states that the last row has no selected content would be useful

apoorvapendse replied:
Updated:
<details>
<summary> diff </summary>

```diff
         const $end = copy_rows.at(-1);
         assert($end?.[0] !== undefined);
         const end_node = the($end.find(".message_content"));
@@ -178,12 +181,13 @@ function construct_copy_div($div: JQuery, start_id: number, end_id: number): voi
         // a single range for a multi-message selection.
         const selected_message_content_elements = get_selected_message_content_elements();
         assert(selected_message_content_elements !== undefined);
+        // Case where the last message doesn't have any highlighted `.message_content`.
+        // Here, end_id is set to id of the message whose username at the top
+        // was highlighted, but has no highlighted `.message_content`.
+        // (See analyze_selection for details.)
+        // So the actually useful/contentful last message of this selection is
+        // at copy_rows[copy_rows.length - 2]
         if (selected_message_content_elements.length === copy_rows.length - 1) {
-            // Case when end_id is set to id of the message whose username at the top
-            // was highlighted, but has no highlighted `.message_content`.
-            // (See analyze_selection for details.)
-            // So the actually useful/contentful last message of this selection is
-            // at copy_rows[copy_rows.length - 2]
             copy_rows.splice(-1, 1);
             if (copy_rows.length === 0) {
                 // In case this just involved selecting the username of a message.
(END)
```
</details>

shubham-padia reviewed (commented):
I've reviewed the first commit and left comments which should resonate with the rest of the PR. We need another pass on this PR to improve the readability of the code. The code should be easy-ish to read for anyone who doesn't understand this system deeply. 

I also had claude do a review of the PR and it highlighted these issues which I verified:
1. Selecting the following gives me a broken list:

<img width="761" height="237" alt="Image" src="https://github.com/user-attachments/assets/ac91629f-ada2-42d8-9bd6-fd0cb84f5db1" />

```
**Imported User:**
Another option is to make the whole bar horizontal and put it on top.* Myrmica ruginodis
* Myrmica kotokui
* Crematogaster pilosa
**Desdemona:**
yello
```

2. Copying the recipient header and the name without the body gives an assertion error:

<img width="989" height="663" alt="Image" src="https://github.com/user-attachments/assets/349363e6-ed42-4940-aa48-4d00bd7f5f21" />

apoorvapendse commented:
> Selecting the following gives me a broken list:


Interesting, thanks for spotting. Looks like quoting is also broken and bleeds into the next selected messages.
I'll patch that. 
I think the easiest solution would be to just insert an extra newline char between messages.

apoorvapendse commented:
fixed some bugs with quotes/list cascading into message content for messages that follow, along with the broken list and assertion error you reported.

<details> 

<summary> screenshots </summary>

<img width="983" height="633" alt="image" src="https://github.com/user-attachments/assets/9b14001a-2d0c-4b69-a4e2-b9d92dcd5309" />


<img width="845" height="750" alt="image" src="https://github.com/user-attachments/assets/b61b3a08-dbce-4354-9775-f1baaa4e4c28" />

</details>

apoorvapendse commented:
Refactored some more, especially the `get_html_for_bookend_message_content`, no more `idx` now.

shubham-padia commented on web/src/copy_messages.ts:
```diff
@@ -48,13 +48,24 @@ function find_boundary_tr(
+        $("<span>").text("| " + date_text),
+    );
+
+    return $p;
```
I'm not a fan of using JS this way to construct html, it becomes hard to read. What do you think about https://sharepatch.com/patches/srE6XMSS7d#diff-file-web-src-copy-messages-ts where `<p data-message-header-paragraph="true"><strong>{{recipient_text}}</strong><span> | {{date_text}}</span></p>
` is `copied_recipient_header.hbs`.

Along with having that template, it doesn't use arithmetic to slice the string (which feels more readable to me, but there can be a similar argument for the current approach as well, so I'll leave it to you to evaluate that). Same with how `recipient_text` is calculated.

apoorvapendse replied:
Yeah that's much cleaner, thank you for the patch.

shubham-padia commented on web/src/compose_paste.ts:
```diff
@@ -629,7 +629,7 @@ export function paste_handler_converter(
 
     // Removes newlines before the start of a list and between list elements.
-    markdown_text = markdown_text.replaceAll(/\n+([*+-])/g, "\n$1");
+    markdown_text = markdown_text.replaceAll(/\n+([*+-])\s+/g, "\n$1 ");
```
Copying the text from https://docs.google.com/document/d/1YrKfF9Ck2gUtrADBmcC4H1STe2s_ZN8arsPwFxeeNQo/edit?usp=sharing results in 

```
**

Section one
- Section two

**
``` 
on this branch

and 

```
**

Section one
-

Section two
**
```

on CZO. Caught by Claude.

apoorvapendse replied:
wow!

apoorvapendse replied:
Updated the regex to avoid eating newlines after the last list item.
```
    markdown_text = markdown_text.replaceAll(/\n+([*+-])[ \t]+/g, "\n$1 ");
```

shubham-padia reviewed (commented):
Left 2 comments, rest looks good!

shubham-padia commented:
Looks good! Merged, thanks @apoorvapendse
